### PR TITLE
auth with github app implementation [closes #1458]

### LINKED ISSUE #1458: [feature] reduce auth scope
## Problem 

https://github.com/inlang/inlang/discussions/1120

Specifically https://github.com/inlang/inlang/discussions/1120#discussioncomment-6597451

## Proposal 

More fine-grained auth without lock-in to Github.

### PR DESCRIPTION
**Pull request by janfjohannes**

### CODE REVIEW

**Comment by felixhaeberle** [comment]
@janfjohannes Can you assign reviewers?

**Comment by janfjohannes** [comment]
explanation how the auth flows behave and look:
https://www.loom.com/share/3cc7d1b35cd84f4c8d1f1987445e1d95?sid=4aae9def-dd4f-419e-9fe0-acee0bbecd49

**Comment by samuelstroschein** [comment]
@janfjohannes can i review this tomorrow?

**Comment by janfjohannes** [comment]
@samuelstroschein  sure, this will not be a blocker until thursday

File: inlang/source-code/website/src/pages/editor/@host/@owner/@repository/State.tsx [hunk 1]
```diff
@@ -196,39 +195,45 @@ export function EditorStateProvider(props: { children: JSXElement }) {
...
+	// get lix errors
+	const [lixErrors, setLixErrors] = createSignal<ReturnType<Repository["errors"]>>([])
+	createEffect(() => {
+		// reset errors to empty on repo changes
+		setLixErrors([])
+
+		repo()?.errors.subscribe((errors) => {
+			setLixErrors(errors)
+		})
+	})
```
Comment (samuelstroschein) [inline inlang/source-code/website/src/pages/editor/@host/@owner/@repository/State.tsx:209]: Why is this needed. Can't you use repo().errors directly where you need it?

File: inlang/source-code/website/src/pages/editor/@host/@owner/@repository/index.page.tsx [hunk 2]
```diff
@@ -71,19 +72,30 @@ function TheActualPage() {
 					</p>
 				}
 			>
-				<Match when={lixErrors().some((e) => e.message.includes("401"))}>
-					<RepositoryDoesNotExistOrNotAuthorizedCard code={401} />
+				<Match when={lixErrors().some((err) => err.message.includes("401"))}>
```
Comment (samuelstroschein) [inline inlang/source-code/website/src/pages/editor/@host/@owner/@repository/index.page.tsx:75]: Using `repo.errors()` seems fitting. Unsure why you need to define lixErrors in state

File: inlang/source-code/website/src/services/auth/src/onSignOut.ts [hunk 3]
```diff
@@ -1,25 +1,13 @@
...
-
-	if (hasExternalGitProxy) {
-		await Promise.allSettled([
-			fetch("/services/auth/sign-out", { method: "POST" }),
-			browserAuth.logout(),
-		])
-	} else {
-		await fetch("/services/auth/sign-out", { method: "POST" })
-	}
+	await browserAuth.logout()
```
Comment (samuelstroschein) [inline inlang/source-code/website/src/services/auth/src/onSignOut.ts:10]: de-coupling auth from the `repo` object seems to be the right design decision :)

File: lix/source-code/client/src/api.ts [hunk 4]
```diff
@@ -9,6 +9,8 @@ type Author = {
 	timezoneOffset?: number
 }
 
+export type LixError = Error & { response?: { status?: number } }
```
Comment (samuelstroschein) [inline lix/source-code/client/src/api.ts:12]: I recommend to extend the Error class to stay compliant with JS and increase interop. Take a look at the errors in `@inlang/sdk`.

```ts
class LixError extends Error {
  // ...
}

```

File: lix/source-code/client/src/browser-auth.ts [hunk 5]
```diff
@@ -95,4 +119,4 @@ export async function getUser() {
 	}
 }
 
-export default { login, logout, getUser }
+export default { login, logout, getUser, addPermissions }
```
Comment (samuelstroschein) [inline lix/source-code/client/src/browser-auth.ts:122]: why export default? in that moment you break tree-shaking. 

i recommend to export each function that should be exported. consumers can then decide what they want to consume, or use a namespace import `import * as auth` to achieve the default export like imports without loosing tree-shaking.

File: lix/source-code/server/src/git-proxy.ts [hunk 6]
```diff
@@ -44,7 +44,7 @@ export async function proxy(request: Request, response: Response, next: NextFunc
 	}
 	try {
 		// remove the proxy path from the url
-		request.url = request.url.split(privateEnv.PUBLIC_GIT_PROXY_PATH)[1]!
+		request.url = request.url.split("/git-proxy/")[1]!
```
Comment (samuelstroschein) [inline lix/source-code/server/src/git-proxy.ts:48]: Removing the null assertion and minimize the risk for bugs. 


```suggestion
		const withoutProxyPath = request.url.split("/git-proxy/")[1]
		if (withoutProxyPath === undefined) {
			throw new Error("Proxy path not found.")
		}
		request.url = withoutProxyPath
```

File: lix/source-code/server/src/github-proxy.ts [hunk 7]
```diff
@@ -48,7 +48,11 @@ router.all(
 			response.set("Access-Control-Allow-Credentials", "true")
 			response.set("Access-Control-Allow-Headers", "x-github-api-version")
 
-			if (res.headers.get("content-type")?.includes("json")) {
+			if (targetUrl!.endsWith("/user/emails") && res.status === 401 && decryptedAccessToken) {
+				response.statusMessage = "token_invalid"
+				response.status(401)
+				response.send("Token invalid")
+			} else if (res.headers.get("content-type")?.includes("json")) {
 				response
```
Comment (samuelstroschein) [inline lix/source-code/server/src/github-proxy.ts:56]: remove the need for null assertions


```suggestion
			response.set("Access-Control-Allow-Headers", "x-github-api-version")

			if (targetUrl.endsWith("/user/emails") && res.status === 401 && decryptedAccessToken) {
				response.statusMessage = "token_invalid"
				response.status(401)
				response.send("Token invalid")
			} else if (res.headers.get("content-type")?.includes("json")) {
				const contentType = res.headers.get("content-type")
				if (contentType === null) {
					throw Error("Content type header not found")
				}
```

**Review by samuelstroschein — CHANGES_REQUESTED** [review]
Looks good, minor suggestions.

File: lix/source-code/server/src/github-proxy.ts (hunk 7 above)
Comment (samuelstroschein) [reply lix/source-code/server/src/github-proxy.ts:56]: not all changes are inclided in the suggestion for whatever reason

File: inlang/source-code/website/src/pages/editor/@host/@owner/@repository/State.tsx (hunk 1 above)
Comment (NiklasBuchfink) [reply inlang/source-code/website/src/pages/editor/@host/@owner/@repository/State.tsx:209]: Then no errors are detected at a later time because the subscription is missing.

**Comment by janfjohannes** [comment]
samuel on loom: "what happens if a user doesn't select all repositories here and gets back to the inlang editor. is a warning shown that "the editor still doesn't has the required permissions. did you select all repositories?""

File: lix/source-code/client/src/api.ts (hunk 4 above)
Comment (felixhaeberle) [reply lix/source-code/client/src/api.ts:12]: Second this

**Review by felixhaeberle — APPROVED** [review]
I second @samuelstroschein review & don't have any further remarks. Good job 👍 Auth gets much more versatile & stable now.

File: inlang/source-code/website/src/pages/editor/@host/@owner/@repository/State.tsx (hunk 1 above)
Comment (samuelstroschein) [reply inlang/source-code/website/src/pages/editor/@host/@owner/@repository/State.tsx:209]: Is repo.errors() not reactive like project.errors()?

Comment (NiklasBuchfink) [reply inlang/source-code/website/src/pages/editor/@host/@owner/@repository/State.tsx:209]: According to @NilsJacobsen: Lix needs to be wrapped in a solid adapter with `from` so that signals can be used. Subscribe works without it. Otherwise a helper function is needed.

File: inlang/source-code/website/src/pages/editor/@host/@owner/@repository/index.page.tsx (hunk 2 above)
Comment (janfjohannes) [reply inlang/source-code/website/src/pages/editor/@host/@owner/@repository/index.page.tsx:75]: @samuelstroschein  lets take the errors discussion out of this pr, as it was not touched here.

File: lix/source-code/client/src/browser-auth.ts (hunk 5 above)
Comment (samuelstroschein) [reply lix/source-code/client/src/browser-auth.ts:122]: In that case, I recommend to export a named object.

```ts
export const browserAuth: AuthObject = {
  login, 
  // ...
}

```

**Pros**

- you can define an `AuthObject` type which must be followed (your default export approach doesn't allow that)

**Cons**

N/A

Comment (samuelstroschein) [reply lix/source-code/client/src/browser-auth.ts:122]: Btw if you resolve the conversation, I get no notification to respond

**Comment by janfjohannes** [comment]
@samuelstroschein  i hope this looks ok to merge now.

File: lix/source-code/client/src/api.ts [hunk 8]
```diff
@@ -9,6 +9,23 @@ type Author = {
 	timezoneOffset?: number
 }
 
+export class LixError extends Error {
+	// we currently mix standard errors with github errors and isomorphic git errors, we will start to transition into a clean lix error class as we replace implementations
+	// the response object is added for transitional compatitbility with github sdk errors
+	response?: { status?: number }
+}
+
+export type lixAuthModule = {
```
Comment (samuelstroschein) [inline lix/source-code/client/src/api.ts:18]: Type capitalized pls. `LixAuthModule`

**Review by samuelstroschein — APPROVED** [review]
